src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/SchemaCache.java

```diff
@@ -1,93 +1,313 @@
+import com.datastax.oss.driver.api.core.metadata.schema.SchemaChangeListener;
+import com.datastax.oss.driver.api.core.metadata.schema.SchemaChangeListenerBase;
+import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata;
+import com.datastax.oss.driver.api.core.session.Session;
 import com.fasterxml.jackson.databind.ObjectMapper;
-import com.github.benmanes.caffeine.cache.Cache;
 import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import edu.umd.cs.findbugs.annotations.NonNull;
 import io.smallrye.mutiny.Uni;
 import io.stargate.sgv2.jsonapi.api.request.RequestContext;
+import io.stargate.sgv2.jsonapi.config.DatabaseType;
 import io.stargate.sgv2.jsonapi.config.OperationsConfig;
+import io.stargate.sgv2.jsonapi.service.cqldriver.CQLSessionCache;
+import io.stargate.sgv2.jsonapi.service.cqldriver.CqlSessionCacheFactory;
+import io.stargate.sgv2.jsonapi.service.cqldriver.CqlSessionFactory;
 import jakarta.enterprise.context.ApplicationScoped;
 import jakarta.inject.Inject;
-import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-/** Caches the vector enabled status for all the namespace in schema */
+/**
+ * Top level entry for caching the keyspaces and tables from the backend db
+ *
+ * <p>IMPORTANT: use {@link #getSchemaChangeListener()} and {@link #getDeactivatedTenantConsumer()}
+ * to get callbacks to evict the cache when the schema changes or a tenant is deactivated. This
+ * should be handled in {@link io.stargate.sgv2.jsonapi.service.cqldriver.CqlSessionCacheFactory}
+ *
+ * <p>TODO: There should be a single level cache of keyspace,table not two levels, it will be easier
+ * to size and manage https://github.com/stargate/data-api/issues/2070
+ */
 @ApplicationScoped
 public class SchemaCache {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SchemaCache.class);
 
-  @Inject QueryExecutor queryExecutor;
+  private final CqlSessionCacheFactory sessionCacheFactory;
+  private final DatabaseType databaseType;
+  private final ObjectMapper objectMapper;
+  private final TableCacheFactory tableCacheFactory;
+  private final OperationsConfig operationsConfig;
 
-  @Inject ObjectMapper objectMapper;
+  /** caching the keyspaces we know about which then have all the tables / collections under them */
+  private final LoadingCache<KeyspaceCacheKey, TableBasedSchemaCache> keyspaceCache;
 
-  @Inject OperationsConfig operationsConfig;
+  @Inject
+  public SchemaCache(
+      CqlSessionCacheFactory sessionCacheFactory,
+      ObjectMapper objectMapper,
+      OperationsConfig operationsConfig) {
+    this(sessionCacheFactory, objectMapper, operationsConfig, TableBasedSchemaCache::new);
+  }
+
+  /**
+   * NOTE: must not use the sessionCacheFactory in the ctor or because it will create a circular
```

Comment:
I assume this is used by (unit) tests as well (`@Inject`) for non-unit-tests. Could then add `@VisibleForTesting` too.